### PR TITLE
Adds a fast message id function and an additional cache for it

### PR DESCRIPTION
**Pull request by blacktemplar**

### CODE REVIEW

File: protocols/gossipsub/src/config.rs [hunk 1]
```diff
@@ -133,9 +134,11 @@ pub struct GossipsubConfig {
...
     /// of the same content from being duplicated.
     ///
-    /// The function takes a `GossipsubMessage` as input and outputs a String to be interpreted as
-    /// the message id.
-    message_id_fn: fn(&GossipsubMessage) -> MessageId,
+    /// The function takes a `GenericGossipsubMessage` as input and outputs a String to be
+    /// interpreted as the message id.
+    message_id_fn: fn(&GenericGossipsubMessage<T>) -> MessageId,
+
+    fast_message_id_fn: Option<fn(&RawGossipsubMessage) -> MessageId>,
```
Comment (AgeManning) [inline protocols/gossipsub/src/config.rs:141]: Can we add some description to this guy

**Review by AgeManning — APPROVED** [review]
Looks good to me. Nice! 
Only one small comment.

File: protocols/gossipsub/src/config.rs (hunk 1 above)
Comment (blacktemplar) [reply protocols/gossipsub/src/config.rs]: Oh forgot that, added a description now.

**Comment by blacktemplar** [comment]
I added now also an own type for fast message ids to make clear that the don't need to be related to message ids and do not have to be interchangeable or something.

**Pull request merged**